adds a function to request a refreshtoken using auth flow

koenhalfwerk opened the pull request:
New-PartnerRefreshToken can consume diffrerent scopes to authenticate to partnercenter or customer's exchange.

using a validate set for scope can be up for debate

``` powershell

$refreshToken = New-PartnerRefreshToken -homeTenantID $homeTenantID -ApplicationId $ApplicationId -scope https://outlook.office365.com/.default  -OnlyRefreshToken

$authBody = @{
    scope         = $scope
    refresh_token = $RefreshToken 
    grant_type    = "refresh_token"
}
$authURL = "https://login.microsoftonline.com/$customerId/oauth2/v2.0/token"
$authResponse = Invoke-RestMethod -Method POST -Uri $authURL -Body $authBody -ErrorAction Stop


$headers = @{authorization  = ('Bearer {0}' -f $authResponse.access_token)}
$users = invoke-restmethod -Uri "https://outlook.office.com/adminApi/beta/$customerID/Mailbox" -Headers $headers

KoenhalfwerkWork committed:
adds a new function

koenhalfwerk commented:
@ili101 I figured out how to use the device authentication, meaning that it waits until you answer, instead of a web service running on a custom socket and no further user interaction is required

``` To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code <chars> to authenticate.

KoenhalfwerkWork committed:
I wish there was a suggestionset instead of validate set

koenhalfwerk commented:
the extremely weird thing (to me) is that the code example in de description does not actually use the partnercenter in any step (it uses office.com) but this is how the ``` New-PartnerAccessToken -module Exchange```  ommand works  (according to fiddler)

KoenhalfwerkWork committed:
write-output is supressed by a variable

ili101 commented:
Sorry it's taking me so long, I'm behind on some stuff.
Nice work with Fiddler, didn't know before it works that seamlessly with PS. 
I added kind of a translation from old module to this one inspired by [Find Azure AD and MSOnline cmdlets in Microsoft Graph PowerShell](https://docs.microsoft.com/en-us/powershell/microsoftgraph/azuread-msoline-cmdlet-map?view=graph-powershell-1.0) here [Cmdlet map.md](https://github.com/ili101/PartnerCustomerCommunity/blob/master/Cmdlet%20map.md#new-partnerrefreshtoken-new-partneraccesstoken-web-app-to-refresh-token), still need to add this command.

I see you decided to essentially split New-PartnerAccessToken to also New-PartnerRefreshToken, it didn't work for me at first but I found it's because of this https://stackoverflow.com/a/57274706/7886965 so maybe this flow less preferable but then more simple.
Maybe we should add a -Flow parameter with options OIDC and DeviceCode as in https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-protocols-oidc and implement OIDC latter with Pode.
-Flow will replace `-UseDeviceAuthentication` (and -ServicePrincipal -UseAuthorizationCode not really sure what's the point of this 2 params / flows, probably don't need them)

ili101 committed:
New-PartnerRefreshToken:
Format consistency and similar to New-PartnerAccessToken.
Added to Cmdlet map.
$Scopes now allows other options.
Added error description regarding "Allow public client flows".
Added 5 seconds sleep to Flow DeviceCode.
Set $ErrorActionPreference = 'Stop' on module globally.
